tracer/src/Datadog.Trace/AppSec/Waf/Encoder.cs

```diff
@@ -286,6 +286,15 @@ DdwafObjectStruct GetStringObject(string value)
                     break;
                 }
 
+                case IEnumerable<KeyValuePair<string, bool>> objDict:
+                {
+                    var count = objDict is ICollection<KeyValuePair<string, bool>> dct ? dct.Count : objDict.Count();
+                    ddwafObjectStruct = ProcessKeyValuePairs(objDict, count, &GetKey1, &GetValue1);
```

Comment:
I'm not sure how to proceed here, because `ProcessKeyValuePairs` needs the count before enumerating, to rent some memory from the UnmanagedMemoryPool:
```
 var childrenCount = !applySafetyLimits || count < WafConstants.MaxContainerSize ? count : WafConstants.MaxContainerSize;
                var childrenFromPool = ObjectStructSize * childrenCount < MaxBytesForMaxStringLength;
                var childrenData = childrenFromPool ? pool.Rent() : Marshal.AllocCoTaskMem(ObjectStructSize * childrenCount);
```
and the memory space is later used in the `foreach`:

```
  foreach (var originalKeyValue in (Dictionary<TKeySource, TValueSource>)enumerableDic)
                        {
                            (...)
                            *(DdwafObjectStruct*)itemData = Encode(getValue(keyValue!), argToFree, applySafetyLimits: applySafetyLimits, key: key, remainingDepth: remainingDepth, pool: pool);
                            itemData += ObjectStructSize;
                        }
 ```
 whereas if we got the children count within the enumeration, we would only get it in the end...
 
 there might still be a way with deeper changes, but like you said, for another PR, I'll create a backlog :)

Reply:
Ah, I see, I misread when `ProcessKeyValuePairs` uses the count. In that case, the following would avoid double enumeration, but would allocate another list. That's probably still preferable IMO as enumerating twice might not even be valid

```
case IEnumerable<KeyValuePair<string, bool>> objDict:
{
    // Create a list if it doesn't implement ICollection
    var collectionDict = objDict is ICollection<KeyValuePair<string, bool>> dct ? dct : dct.ToList();
    var count = collectionDict.Count;
    ddwafObjectStruct = ProcessKeyValuePairs(collectionDict, count, &GetKey1, &GetValue1);
```

Definitely one to look into in a different PR tho

Reply:
thanks, temp fix suggested added in [9106f1b](https://github.com/DataDog/dd-trace-dotnet/pull/4654/commits/9106f1bd24b822af09dc1b1bd99fa390ad184a84)